heron/uploaders/src/java/com/twitter/heron/uploader/gcs/GcsUploader.java

```
@@ -0,0 +1,225 @@
+      } catch (IOException ioe) {
+        throw new UploaderException("Failed to back up previous topology version", ioe);
+      }
+    }
+
+    final StorageObject storageObject;
+    try {
+      storageObject = gcsController.createObject(topologyObjectName, topologyPackageFile);
+    } catch (IOException ioe) {
+      throw new UploaderException(
+          String.format("Error writing topology package to %s %s",
+              bucket, topologyObjectName), ioe);
+    }
+
+    final String downloadUrl = getDownloadUrl(bucket, storageObject.getName());
+    LOG.info("Package URL: " + downloadUrl);
+    try {
+      return new URI(downloadUrl);
+    } catch (URISyntaxException e) {
+      throw new UploaderException(
+          String.format("Could not convert URL %s to URI", downloadUrl), e);
+    }
+  }
+
+  @Override
+  public boolean undo() {
+    // Try to get the previous version. This will be null on the first deploy.
+    final StorageObject previousObject = gcsController.getObject(previousTopologyObjectName);
+    if (previousObject != null) {
+      try {
+        gcsController.copyObject(previousTopologyObjectName, topologyObjectName, previousObject);
+      } catch (IOException ioe) {
+        LOG.log(Level.SEVERE, "Reverting to previous topology version failed", ioe);
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  @Override
+  public void close() {
+    if (!Strings.isNullOrEmpty(previousTopologyObjectName)) {
+      try {
+        gcsController.deleteObject(previousTopologyObjectName);
+      } catch (IOException ioe) {
+        LOG.info("Failed to delete previous topology " + previousTopologyObjectName);
+      }
+    }
+  }
+
+  GcsController createGcsController(Config configuration, String storageBucket) {
+    final Credential credential;
+    try {
+      credential = createCredentials(configuration);
+    } catch (IOException ioe) {
+      throw new RuntimeException("Unable to create credentials", ioe);
```

Here and elsewhere can we throw the original exception and not wrap in RuntimeException?